Added helper for traits-compatible getDeclaringClass

enumag opened the pull request:
The `ReflectionProperty::getDeclaringClass()` doesn't work with traits, it instead always returns the class which used the trait, not the actual trait that declared the property. The same goes for ReflectionMethod as well.

This affects for example the `@inject` annotations from Nette\DI and the `@autowire` annotations from Kdyby\Autowired.

**Example:**
```php
// FooTrait.php
namespace Foo;
use Foo\Model\Service;
trait FooTrait {
	/** @var Service @inject */
	public $service;
}

// FooPresenter.php
namespace Foo;
class FooPresenter extends \Nette\Application\UI\Presenter
{
	use FooTrait;
}
```

*With such code the DI container won't be able to inject the Foo\Model\Service class because it will look to the use statements in FooPresenter.php and look for class Foo\Service instead.*

**Limitations of the solution:**
- When the property is declared in multiple traits, the PHP emits some kind of warning. If this warning is ignored, this won't work correctly.
- This solution wouldn't work for methods renamed using `use FooTrait { method as newMethod }`.

**TODO:**
- Tests.
- Add the same for methods.
- Discuss the class & methods names & parameters.
- Discuss possible integration into `AnnotationsParser::expandClassName()` - it could accept ReflectionMethod and ReflectionProperty as well and search for the class/trait using this.

**Related links:**
- https://github.com/Kdyby/Autowired/issues/17
- http://mouf-php.com/blog/php_reflection_api_traits

@dg I'd like to hear your thoughts about this before working on it further. Mainly whether you want such solution in Nette at all.

dg commented:
:+1: for adding

enumag commented:
@dg Good. :-)

What do you think about the integration into `AnnotationsParser::expandClassName()`?

Now for the API. My personal preference is this:

```php
/**
 * @param \ReflectionProperty|\ReflectionMethod
 * @return Nette\Reflection\ClassType
 */ 
public static function getDeclaringClass(\Reflector $reflection);
```

Possibly we can go with something like this:
```php
// all parameters strings or ReflectionClass $class and string $property/$method
public static function getPropertyDeclaringClass($class, $property);
public static function getMethodDeclaringClass($class, $method);
```

Also I'm not completely sure about the class name TraitHelpers. If anyone has a better idea, please tell me. :-)

dg commented:
Simply Helpers.

JanTvrdik commented on src/Reflection/Helpers.php:
```diff
@@ -0,0 +1,69 @@
+
+			} elseif ($mode === self::MODE_PROPERTY && $trait->hasProperty($name) || $mode === self::MODE_METHOD && $trait->hasMethod($name)) {
+				return $trait;
+			}
```
What about

~~~php
} elseif ($trait->$method($name)) {
    return $trait;
}
~~~
?

enumag replied:
Actually this will have to be done otherwise. Wait a moment please.

JanTvrdik commented on src/Reflection/Helpers.php:
```diff
@@ -0,0 +1,69 @@
+			return new ClassType($trait->getName());
+		}
+	
+		return new ClassType($class->getName());
```
~~~php
return new ClassType(($trait ?: $class)->getName());
~~~

enumag commented:
@JanTvrdik All done... hopefully... Please review now. :-)

enumag commented:
Hmm... How should I update the test to run on PHP 5.4+ only? The traits cause parse error in 5.3 of course.

dg commented:
Why are you checking for line numbers?

enumag commented:
Because it wouldn't work for multiple traits in one file.

JanTvrdik commented:
@enumag Which seems like too insane use-case to waste lines of code to support it.

dg commented:
I still don't understand. There is hasMethod(). Filename&line based solution seems crazy.

enumag commented:
Unlike properties, methods can be overriden without E_STRICT warning. That's why hasMethod is not enough and I need to check the file name as well. I can drop the line numbers but I'd have to split the test classes into multiple files because the tests would fail otherwise. This seems more clean (both the code and the tests).

enumag commented:
Or I can just drop the support for methods entirely. AFAIK we only need this for properties anyway (for now).

JanTvrdik commented:
:+1: for dropping the support for method if we don't need it but I think that you don't need to check the line numbers, you just have to fix the order in which you traverse the class tree.

enumag commented:
@JanTvrdik Ok, I'll drop it then. We can add it later if needed.

hrach commented:
using "we need" in nette/reflection package in context of nette/di packagace is quite dangerous.

dg commented:
Thanx for explanation. Ok, in this case it must be filename&line based. Can you add test that uses `insteadof` or `as`?

enumag commented:
@dg As I've written in the first comment, that is the limitation of this solution. It won't work.

enumag commented:
Afaik `insteadof` and `as` don't work for properties so it should be ok. I've dropped the methods support and simplified the remaining code according to @JanTvrdik 's suggestions (thanks!).

dg commented:
Why is `insteadof` or `as` limitation when you are checking filenames & lines?

enumag commented:
Sorry, my fault. That was the case with different API (when I'd only have the method's name, not ReflectionMethod object).

Does that mean you want the support for methods?

JanTvrdik commented:
@dg Do we have a use-case the the methods?

dg commented:
Oh… http://3v4l.org/hiJag

dg commented:
@JanTvrdik DI uses only annotation `@return`, so use-case is only hypothetical.

enumag commented:
@dg Actually it uses `@param` annotations as well. Only for constructors and interfaces though.

dg commented:
Really? Where is it in code?

enumag commented:
Hmpf, nowhere.

dg commented:
Merged. I little bit changed implementation.

enumag commented:
FIne, thx.
